src/BuiltInTools/BrowserRefresh/BlazorWasmHotReloadMiddleware.cs

```diff
@@ -0,0 +1,140 @@
+
+namespace Microsoft.AspNetCore.Watch.BrowserRefresh
+{
+    /// <summary>
+    /// A middleware that manages receiving and sending deltas from a BlazorWebAssembly app.
+    /// This assembly is shared between Visual Studio and dotnet-watch. By putting some of the complexity
+    /// in here, we can avoid duplicating work in watch and VS.
+    /// </summary>
+    internal sealed class BlazorWasmHotReloadMiddleware
+    {
+        private readonly object @lock = new();
+        private readonly string EtagDiscriminator = Guid.NewGuid().ToString();
+        private readonly JsonSerializerOptions _jsonSerializerOptions = new()
+        {
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        };
+
+        public BlazorWasmHotReloadMiddleware(RequestDelegate next)
+        {
+        }
+
+        internal List<UpdateDelta> Deltas { get; } = new();
+
+        public Task InvokeAsync(HttpContext context)
+        {
+            // Multiple instances of the BlazorWebAssembly app could be running (multiple tabs or multiple browsers).
+            // We want to avoid serialize reads and writes between then
+            lock (@lock)
+            {
+                if (HttpMethods.IsGet(context.Request.Method))
+                {
+                    return OnGet(context);
+                }
+                else if (HttpMethods.IsPost(context.Request.Method))
+                {
+                    return OnPost(context);
+                }
+                else
+                {
+                    context.Response.StatusCode = StatusCodes.Status405MethodNotAllowed;
+                    return Task.CompletedTask;
+                }
+            }
+
+            // Don't call next(). This middleware is terminal.
+        }
+
+        private async Task OnGet(HttpContext context)
+        {
+            if (Deltas.Count == 0)
+            {
+                context.Response.StatusCode = StatusCodes.Status204NoContent;
+                return;
+            }
+
+            if (EtagMatches(context))
+            {
+                context.Response.StatusCode = StatusCodes.Status304NotModified;
+                return;
+            }
+
+            WriteETag(context);
+            await JsonSerializer.SerializeAsync(context.Response.Body, Deltas, _jsonSerializerOptions);
+        }
+
+        private bool EtagMatches(HttpContext context)
+        {
+            if (context.Request.Headers[HeaderNames.IfNoneMatch] is not { Count: 1 } ifNoneMatch)
+            {
+                return false;
+            }
+
+            var expected = GetETag();
+            return string.Equals(expected, ifNoneMatch[0], StringComparison.Ordinal);
+        }
+
+        private async Task OnPost(HttpContext context)
+        {
+            var updateDeltas = await JsonSerializer.DeserializeAsync<UpdateDelta[]>(context.Request.Body, _jsonSerializerOptions);
+            AppendDeltas(updateDeltas);
+
+            WriteETag(context);
```

Comment:
This is called by VS or `dotnet watch`? Could you add some comments at specific places to help explain the flow to folks newer to this area how it works? (Or better yet write a design issue?)